rclcpp/include/rclcpp/experimental/buffers/ring_buffer_implementation.hpp

```diff
@@ -88,16 +88,19 @@ class RingBufferImplementation : public BufferImplementationBase<BufferT>
 
   inline size_t next(size_t val)
   {
+    std::lock_guard<std::recursive_mutex> lock(recursive_mutex_);
```

Comment:
I think that methods for `next`, `has_data` and `is_full` can be implemented via reader lock, using std::shared_mutex instead of std::mutex.

Reply:
I think you're right.  I'll give this a try.

Reply:
I tried using a `shared_mutex` with a `unique_lock` (also tried `lock_guard`) in the write functions (i.e., `enqueue()` and `dequeue()`) and `shared_lock`s in the read functions (i.e., `next()`, `has_data()`, and `is_full()`).  This doesn't work because these read operations are called in the write functions, causing a deadlock.

Any ideas on how to use a `shared_mutex` on the read functions, considering that the read functions are called in the write functions?

Reply:
basically I think that mutex should be aligned with data but function.

how about the following?

- `enqueue` w/o `next` / `is_full` and use `unique_lock`
- `dequeue` w/o `has_data` / `next` and use `unique_lock`
- `next` with `shared_lock`
- `has_data` with `shared_lock`
- `is_full` with `shared_lock` 

with above,

- it can keep the interfaces compatible.
- it can use reader/writer lock instead of mutex. (also no need for recursive mutex)

I think that we should avoid the complicated code with mutex, but since this is the place for performance, I'd think using shared_mutex would be better.